### ISSUE TITLE
numpy.linalg.LinAlgError: Eigenvalues did not converge on ARM64 builds

### ISSUE THREAD

**Issue by ogrisel** [numpy/numpy#19411]
`TestPolynomial.test_poly` and other tests recently started to fail on the Travis ARM64 nightly builds:

https://travis-ci.com/github/numpy/numpy/jobs/521809468

The last successful ARM64 test run is 4 days old:

https://travis-ci.com/github/numpy/numpy/builds/231467572

Note that we observed similar failures on Circle CI and Travis CI ARM64 builds for the scikit-learn project using only stable releases for numpy and scipy so this probably not caused by a change in numpy itself.

We observed a failing job and a successful run that both used:

- numpy-1.21.0-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
- scipy-1.7.0-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl

Link to those (weekly running) scikit-learn jobs:

- first failure: https://travis-ci.com/github/scikit-learn/scikit-learn/jobs/521594473
- recent success: https://travis-ci.com/github/scikit-learn/scikit-learn/jobs/519470685

I am not sure what caused this change since openblas is embedded in those wheels.

**Comment by charris** [comment]
There are some passing tests. Looking at the worker info at the top of  the jobs:

- lxd-arm64-02 fails
- lxd-arm64-05 works
- lxd-arm64-06 works

The 1.21.x branch has also started failing, so I don't think the problem is in numpy.

**Comment by ogrisel** [comment]
This is really weird. Any idea how to find the cause of this problem? It seems that there was no change in numpy or openblas that could explain this. Could it be some faulty hardware or another change in the software environment?

I also tried to run those tests locally on a linux arm64 docker image on an Apple M1 machine they pass without any problem.

**Comment by mattip** [comment]
A successful job reports [build system information](https://travis-ci.com/github/scikit-learn/scikit-learn/jobs/519470685#L20)
```
Runtime kernel version: 5.8.0-55-generic
travis-build version: 091d532a
Build image provisioning date and time
Fri Jul 10 13:20:51 UTC 2020
Operating System Details
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.6 LTS
Release:	16.04
Codename:	xenial
Linux Version
5.0.0-23-generic
```

A failing job [reports](https://travis-ci.com/github/numpy/numpy/jobs/521809468#L14)
```
Runtime kernel version: 5.8.0-59-generic
travis-build version: 091d532a
Build image provisioning date and time
Mon Nov  2 09:23:45 UTC 2020
Operating System Details
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.1 LTS
Release:	20.04
Codename:	focal
Linux Version
5.0.0-23-generic
```
```

**Comment by mattip** [comment]
I opened an issue on the travis-ci forum, please comment/upvote so it gets some attention
https://travis-ci.community/t/arm64-builds-failing-blas-tests-did-the-hardware-change-recently/11813

**Comment by ogrisel** [comment]
On my local machine with a successful run:

```
$ uname -a
Linux a8c27e8c4c7a 5.10.25-linuxkit #1 SMP PREEMPT Tue Mar 23 09:24:45 UTC 2021 aarch64 GNU/Linux
$ cat /etc/issue.net
Debian GNU/Linux 10
$ cat /etc/debian_version 
10.7
$ lspcu
Architecture:        aarch64
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
Vendor ID:           ARM
Model:               0
Model name:          Cortex-A57
Stepping:            r1p0
BogoMIPS:            48.00
Flags:               fp asimd evtstrm aes pmull sha1 sha2 crc32 fphp asimdhp cpuid dit
```

**Comment by ogrisel** [comment]
We might want to run `lscpu` before running the tests to collect additional info.

**Comment by ogrisel** [comment]
Note that this is not travis specific: we also observe the same problem on circle ci:

https://app.circleci.com/pipelines/github/scikit-learn/scikit-learn/15376/workflows/6dc4ec4f-c32a-4bc8-be42-a6ac42bb346f/jobs/143767

But maybe the use the same hardware provider as travis (travis uses https://metal.equinix.com/, but I don't know for circle ci).

**Comment by Qiyu8** [comment]
My centOS-based local machine(KunPeng 920) can run successfully.  but there has an compiler error when parallel build is enabled.
```
$ uname -a
Linux ecs-9d50 4.19.36-vhulk1907.1.0.h962.eulerosv2r8.aarch64 #1 SMP Fri Jan 8 13:18:01 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
$ cat /etc/os-release
NAME="EulerOS"
VERSION="2.0 (SP8)"
ID="euleros"
ID_LIKE="rhel fedora centos"
VERSION_ID="2.0"
PRETTY_NAME="EulerOS 2.0 (SP8)"
ANSI_COLOR="0;31"
$ lscpu
Architecture:        aarch64
Byte Order:          Little Endian
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  1
Core(s) per socket:  8
Socket(s):           1
NUMA node(s):        1
Vendor ID:           0x48
Model:               0
Stepping:            0x1
CPU MHz:             2600.000
CPU max MHz:         2600.0000
CPU min MHz:         2600.0000
BogoMIPS:            200.00
L1d cache:           64K
L1i cache:           64K
L2 cache:            512K
L3 cache:            32768K
NUMA node0 CPU(s):   0-7
Flags:               fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma dcpop asimddp asimdfhm
```

**Comment by ogrisel** [comment]
Here is the lscpu output of a [failed run](https://app.travis-ci.com/github/scikit-learn/scikit-learn/builds/231938988) on travis:

```
Architecture:          aarch64
Byte Order:            Little Endian
CPU(s):                80
On-line CPU(s) list:   0-79
Thread(s) per core:    1
Core(s) per socket:    80
Socket(s):             1
NUMA node(s):          1
L1d cache:             64K
L1i cache:             64K
L2 cache:              1024K
NUMA node0 CPU(s):     0-79
```

Unfortunately, this did not print the CPU flags. Maybe because of virtualization?

**Comment by ogrisel** [comment]
I am also tried to trigger a run on conda-forge and the [scikit-learn tests pass](https://cloud.drone.io/conda-forge/scikit-learn-feedstock/142/6/2) on the ARM64 workers of the drone CI.

**Comment by mattip** [comment]
Apparently we can request aarch64 access from https://github.com/WorksOnArm/cluster by [opening a new issue](https://github.com/WorksOnArm/cluster/issues/new/choose) like this one for conda-forge WorksOnArm/cluster#193. I think each project needs to do that separately. From the conda-forge issue it seems that conda-forge, in Feb 2020, got at least a "c2.large.arm machine", which seem to be [these offerings](https://metal.equinix.com/developers/docs/servers/server-specs/#arm-servers) from Equinix Metal Server. For NumPy, let's discuss this at the weekly meeting.

**Comment by ogrisel** [comment]
I think I am making some progress: on the travis machine where I observe the failure, the one that has not `flags` entry in the output of `lscpu`, OpenBLAS detects the `neoversen1` instead of the `armv8` on my Apple M1 machine under macOS and `cortexa57` on the same machine under linux aarch64 via docker. The tests pass on my machine both with `cortexa57` and `armv8`.

To introspect the corename detected by openblas, use the following augmented version of threadpoolctl (see https://github.com/joblib/threadpoolctl/pull/85 for details):

```
$ pip install git+https://github.com/ogrisel/threadpoolctl.git@openblas_get_corename
$ python -m threadpoolctl -i numpy
[
  {
    "filepath": "/home/travis/miniconda/envs/testenv/lib/libopenblasp-r0.3.15.so",
    "prefix": "libopenblas",
    "user_api": "blas",
    "internal_api": "openblas",
    "version": "0.3.15",
    "num_threads": 2,
    "threading_layer": "pthreads",
    "corename": "neoversen1"
  }
]
```

Here is the (failing) run:

https://app.travis-ci.com/github/scikit-learn/scikit-learn/builds/232084976#L3799-L3825


If I force the use of the use of the `armv8` by setting the following env variable:

```
export OPENBLAS_CORETYPE=armv8
```

then the tests pass again on travis:

https://app.travis-ci.com/github/scikit-learn/scikit-learn/builds/232087064


So it's possible that the problem comes from a bug in the `neoversen1` kernel in OpenBLAS 0.3.15 (latest version at the moment) or that the travis machines make OpenBLAS detect the wrong kind of kernel (because the CPU flags are not properly reported).

**Comment by ogrisel** [comment]
So a workaround for now it force `export OPENBLAS_CORETYPE=armv8` in the Travis CI config of numpy, scipy, scikit-learn...

**Comment by charris** [comment]
Using `arch: arm64-graviton2` in `travis.yml` also fixes the problem for NumPy.  See #19426. Note s390x seems to have gone missing.

**Comment by ogrisel** [comment]
For information, OpenBLAS detects the ARM64 variants based on the "cpu part" element of the contents of `/proc/cpuinfo` (see the [source code](https://github.com/xianyi/OpenBLAS/blob/1a8b6134c24f66c649ca342c74d551055b98cbc6/cpuid_arm64.c#L165-L166) for more details):

On the failing travis node, the contents of `/proc/cpuinfo` is:

```
processor	: 0
BogoMIPS	: 50.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x3
CPU part	: 0xd0c
CPU revision	: 1

processor	: 1
BogoMIPS	: 50.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x3
CPU part	: 0xd0c
CPU revision	: 1

processor	: 2
BogoMIPS	: 50.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x3
CPU part	: 0xd0c
CPU revision	: 1
[...]
```
I will try to do a proper bug report to OpenBLAS developers tomorrow.

**Comment by charris** [comment]
Unfortunately, the fixes don't work for the docker images used to test the numpy wheels.

**Comment by ogrisel** [comment]
Can you please run:

```
lscpu
pip install git+https://github.com/ogrisel/threadpoolctl.git@openblas_get_corename
python -m threadpoolctl -i numpy
cat /proc/cpuinfo
```

before running the tests on the numpy wheels CI?

**Comment by charris** [comment]
@ogrisel 

_graviton2_
```
[
  {
    "filepath": "/venv/lib/python3.7/site-packages/numpy.libs/libopenblasp-r0-afb71072.3.13.dev.so",
    "prefix": "libopenblas",
    "user_api": "blas",
    "internal_api": "openblas",
    "version": "0.3.13.dev",
    "num_threads": 2,
    "threading_layer": "pthreads",
    "corename": "neoversen1"
  }
]
processor	: 0
BogoMIPS	: 243.75
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x3
CPU part	: 0xd0c
CPU revision	: 1

processor	: 1
BogoMIPS	: 243.75
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x3
CPU part	: 0xd0c
CPU revision	: 1
```

arm64 (80 processors)
```
[
  {
    "filepath": "/venv/lib/python3.8/site-packages/numpy.libs/libopenblasp-r0-afb71072.3.13.dev.so",
    "prefix": "libopenblas",
    "user_api": "blas",
    "internal_api": "openblas",
    "version": "0.3.13.dev",
    "num_threads": 64,
    "threading_layer": "pthreads",
    "corename": "neoversen1"
  }
]
processor	: 0
BogoMIPS	: 50.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x3
CPU part	: 0xd0c
CPU revision	: 1
```

**Comment by ogrisel** [comment]
So indeed the `arm64 (80 processors)` and `graviton2` look very similar from an lscpu point of view (in particular the CPU part identifier than OpenBLAS relies on to detect the CPU architecture is the same).

> Unfortunately, the fixes don't work for the docker images used to test the numpy wheels.

What do you mean? Setting the `OPENBLAS_CORETYPE=armv8` environment variable does not work on the `arm64` host used to generate the manylinux wheels? Or using the graviton2 host does not work for the manylinux wheels?

In any case, we should probably try to write a minimal C reproducer to report the issue to OpenBLAS but I won't have the time to do it soon...

**Comment by ogrisel** [comment]
It seems that the problem has disappeared magically on the scikit-learn Travis CI (see https://github.com/scikit-learn/scikit-learn/pull/20476 after `a1e1a03`). There is still a failure but it's unrelated to the originally reported `numpy.linalg.LinAlgError: Eigenvalues did not converge` problem.  Here are the CPU info on the arm64 80 cores machine of the last successful run.

```
[
  {
    "filepath": "/home/travis/miniconda/envs/testenv/lib/libgomp.so.1.0.0",
    "prefix": "libgomp",
    "user_api": "openmp",
    "internal_api": "openmp",
    "version": null,
    "num_threads": 2
  },
  {
    "filepath": "/home/travis/miniconda/envs/testenv/lib/libopenblasp-r0.3.17.so",
    "prefix": "libopenblas",
    "user_api": "blas",
    "internal_api": "openblas",
    "version": "0.3.17",
    "num_threads": 2,
    "threading_layer": "pthreads",
    "architecture": "neoversen1"
  }
]
processor	: 0
BogoMIPS	: 50.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x3
CPU part	: 0xd0c
CPU revision	: 1
```

It seems the be the same as for the [previously failing runs](https://github.com/numpy/numpy/issues/19411#issuecomment-875966514)... so I am confused. Maybe this was a transient hardware / virtualization problem on the hosting provider...

In anycase we can no longer debug. Maybe we can close.

**Comment by charris** [comment]
There was an OpenBLAS fix for the problem in 0.3.16, probably https://github.com/xianyi/OpenBLAS/pull/3278.

**Comment by ogrisel** [comment]
> There was an OpenBLAS fix for the problem in 0.3.16, probably xianyi/OpenBLAS#3278.

It seems unrelated (not the same CPU part: `0xd05` vs `0xd0c` on Travis CI) however it's true that now the new runs are running with OpenBLAS 0.3.17 while the failed run used: 0.3.17.

**Comment by ogrisel** [comment]
The fix was probably released in 0.3.16, from the [changelog](https://github.com/xianyi/OpenBLAS/blob/1b6db3dbba672b4f8af935bd43a1ff6cff4d20b7/Changelog.txt#L56-L58):

> - fixed missing restore of a register in the recently rewritten DNRM2 kernel
  for ThunderX2 and Neoverse N1 that could cause spurious failures in e.g.
  DGEEV
